File: adapters/telemetry/packed.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Crystal Warden Supply Chain Labs LLC
"""Word-packed wire — the real byte format (Phase A's codec is a bit-string reference; this is what goes
on the link).

The Fibonacci bits are accumulated MSB-first into a 64-bit word register and flushed per word; the final
partial word is zero-padded on the right. Because a code ends in a unique ``11`` and the pad is a run of
0s, the padding is a partial frame with no terminator — ``decode_stream`` drops it — so the round-trip is
exact without carrying a bit count. (This is the doc's CPU/MCU path: accumulate-into-a-word for ~10-20x
throughput on C/hardware; the Python reference here nails the *format* + correctness + padding cost, not
the raw speed — that win lives in the C/FPGA codec, Phase C2.)
"""
from __future__ import annotations

from typing import Dict, List

import zeckendorf as z

WORD_BITS = 64


def pack(bits: str, word_bits: int = WORD_BITS) -> bytes:
    """Bit-string -> bytes, MSB-first, flushing per `word_bits`-bit word; final word zero-padded right."""
    out = bytearray()
    acc = 0
    n = 0
    for ch in bits:
        acc = (acc << 1) | (1 if ch == "1" else 0)
        n += 1
        if n == word_bits:
            out += acc.to_bytes(word_bits // 8, "big")
            acc, n = 0, 0
    if n:
        acc <<= (word_bits - n)                      # left-align the partial word; low bits = zero pad
        out += acc.to_bytes(word_bits // 8, "big")
    return bytes(out)


def unpack(data: bytes) -> str:
    """Bytes -> bit-string, MSB-first (includes any trailing zero pad — harmless, dropped on decode)."""
    return "".join(format(b, "08b") for b in data)


def encode(ints: List[int], word_bits: int = WORD_BITS) -> bytes:
    """Positive ints -> Fibonacci-coded, word-packed bytes."""
    return pack(z.encode_stream(ints), word_bits)


def decode(data: bytes) -> List[int]:
    """Word-packed bytes -> ints (trailing zero pad is a partial frame and is dropped)."""
    return z.decode_stream(unpack(data))


def padding_overhead(ints: List[int], word_bits: int = WORD_BITS) -> Dict:
    """Measure the word-padding cost: only the final word is padded, so it amortizes away as N grows."""
    actual_bits = len(z.encode_stream(ints))
    wire = encode(ints, word_bits)
    padded_bits = len(wire) * 8
    pad_bits = padded_bits - actual_bits
    return {"n": len(ints), "actual_bits": actual_bits, "wire_bytes": len(wire),
            "pad_bits": pad_bits, "pad_pct": (100.0 * pad_bits / actual_bits) if actual_bits else 0.0}


if __name__ == "__main__":                           # human-readable padding-overhead table
    print(f"{'N':>8} {'wire_bytes':>11} {'pad_bits':>9} {'pad_%':>7}")
    for n in (10, 100, 1_000, 10_000, 100_000):
        o = padding_overhead(list(range(1, n + 1)))
        print(f"{o['n']:>8} {o['wire_bytes']:>11} {o['pad_bits']:>9} {o['pad_pct']:>6.3f}%")

File: prismpath-telemetry-rs/src/packed.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Crystal Warden Supply Chain Labs LLC
//! Word-packed wire — the real byte format.

use crate::zeckendorf;
use serde::{Deserialize, Serialize};

pub const WORD_BITS: usize = 64;

/// Bit-string -> bytes, MSB-first, flushing per `word_bits`-bit word; final word zero-padded right.
pub fn pack(bits: &str, word_bits: usize) -> Vec<u8> {
    let bytes_per_word = word_bits / 8;
    let mut out = Vec::new();
    let mut acc: u128 = 0;
    let mut n = 0;
    for ch in bits.chars() {
        let bit = if ch == '1' { 1 } else { 0 };
        acc = (acc << 1) | bit;
        n += 1;
        if n == word_bits {
            let be_bytes = acc.to_be_bytes();
            out.extend_from_slice(&be_bytes[16 - bytes_per_word..]);
            acc = 0;
            n = 0;
        }
    }
    if n > 0 {
        acc <<= word_bits - n;
        let be_bytes = acc.to_be_bytes();
        out.extend_from_slice(&be_bytes[16 - bytes_per_word..]);
    }
    out
}

/// Bytes -> bit-string, MSB-first (includes any trailing zero pad).
pub fn unpack(data: &[u8]) -> String {
    data.iter().map(|b| format!("{:08b}", b)).collect()
}

/// Positive ints -> Fibonacci-coded, word-packed bytes. `Err` if any int is `< 1`.
pub fn encode(ints: &[usize], word_bits: usize) -> Result<Vec<u8>, String> {
    Ok(pack(&zeckendorf::encode_stream(ints)?, word_bits))
}

/// Word-packed bytes -> ints (trailing zero pad is a partial frame and is dropped).
pub fn decode(data: &[u8]) -> Vec<usize> {
    zeckendorf::decode_stream(&unpack(data))
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PaddingOverhead {
    pub n: usize,
    pub actual_bits: usize,
    pub wire_bytes: usize,
    pub pad_bits: usize,
    pub pad_pct: f64,
}

pub fn padding_overhead(ints: &[usize], word_bits: usize) -> Result<PaddingOverhead, String> {
    let actual_bits = zeckendorf::encode_stream(ints)?.len();
    let wire = encode(ints, word_bits)?;
    let padded_bits = wire.len() * 8;
    let pad_bits = padded_bits - actual_bits;
    let pad_pct = if actual_bits > 0 {
        (100.0 * pad_bits as f64) / actual_bits as f64
    } else {
        0.0
    };
    Ok(PaddingOverhead {
        n: ints.len(),
        actual_bits,
        wire_bytes: wire.len(),
        pad_bits,
        pad_pct,
    })
}
